File: Interface2App/Flicker.cs

using System;
using System.CodeDom;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Security.Cryptography.X509Certificates;

namespace Interface2App
{

	public class Flicker : EntityValidator
	{
        [Required, RegularExpression(@"^.*", ErrorMessage = "Please enter a number")]
        public string Name { get; set; }
        [Required, RegularExpression(@"^.*[0-9]", ErrorMessage = "Please enter a number")]
		//[Range(0, sizeX())]
		public int X { get; set; }
		[Required, RegularExpression(@"^.*[0-9]", ErrorMessage = "Please enter a number")]
		//[Range(0, 720)]
		public int Y { get; set; }
		[Required, RegularExpression(@"^.*[0-9]", ErrorMessage = "Please enter a number")]
		//[Range(0, 1280)]
		public int Width { get; set; }
		[Required, RegularExpression(@"^.*[0-9]", ErrorMessage = "Please enter a number")]
		//[Range(0, 720)]
		public int Height { get; set; }
		[Required, RegularExpression(@"^.*[0-9]", ErrorMessage = "Please enter a number")]
		public double Frequency { get; set; }
		[Required, RegularExpression(@"^.*[0-9]", ErrorMessage = "Please enter a number")]
		[Range(0, 360)]
		public double Phase { get; set; }
		[Required, RegularExpression(@"^.*[0-9]", ErrorMessage = "Please enter a number")]
		[Range(0, 100)]
		public int Opacity_Min { get; set; }
		public int Opacity_Max { get; set; }
		public Signal_Type Type { get; set; }
        public System.Windows.Media.Color color1 { get; set; }
        public string image { get; set; }
		public bool IsImageFlicker { get; set; } 
		public sequenceValue sequence { get; set; }



        public enum Signal_Type
		{
			Sine=1,
			Root_Square=3,
			Square=2,
			Random=0,
			Maximum_Lenght_Sequence=4,
			None=5
		}
		/// <summary>
		/// create a new flicker in top-left point of the screen with white color.
		/// </summary>
		public Flicker()
		{
			Name = "Flicker";
			Opacity_Min = 0;
			Opacity_Max = 100;
			Type= Signal_Type.Sine;
			color1= System.Windows.Media.Color.FromArgb(255,255,255,255);
			X = 0;
			Y = 0;
			Width = 100;
			Height = 100;
			IsImageFlicker = false;
			image = String.Empty;
			Frequency= 1;
			Phase = 0;
			sequence = new sequenceValue(sequenceValue.type.Block, sequenceValue.CondType.Never);
            sequence.addSeq(new sequenceValue(sequenceValue.type.Active, sequenceValue.CondType.Never));
        }
		/// <summary>
		/// custom flicker with parameters
		/// </summary>
		/// <param name="X"></param>
		/// <param name="Y"></param>
		/// <param name="Width"></param>
		/// <param name="Height"></param>
		/// <param name="c"></param>
        public Flicker(String name,int X,int Y,int Width,int Height,System.Windows.Media.Color c,int op_min,int op_max,double freq,double phase,int type,sequenceValue seq)
        {
            Name = name;
            Opacity_Min = op_min;
            Opacity_Max = op_max;
            Type = (Signal_Type)type;
			color1 = c;
            this.X = X;
            this.Y = Y;
            this.Width = Width;
            this.Height = Height;
            IsImageFlicker = false;
            image = String.Empty;
            Frequency = freq;
			Phase=phase;
			sequence = seq;
        }
        //Other construction method
		public Flicker(int X, int Y, int Width, int Height, System.Windows.Media.Color c, int op_min, int op_max, double freq, double phase, int type) : this("Flicker", X, Y, Width, Height, c, op_min, op_max, freq, phase, type, new sequenceValue(sequenceValue.type.Block, sequenceValue.CondType.Never).addSeq(new sequenceValue(sequenceValue.type.Active, sequenceValue.CondType.Never)))
        {
        }
        public Flicker(string name,int X, int Y, int Width, int Height, System.Windows.Media.Color c, int op_min, int op_max, double freq, double phase, int type) :this(name, X, Y, Width, Height, c, op_min, op_max, freq, phase, type, new sequenceValue(sequenceValue.type.Block, sequenceValue.CondType.Never).addSeq(new sequenceValue(sequenceValue.type.Active, sequenceValue.CondType.Never)))
        {
        }
        public Flicker(int X, int Y, int Width, int Height, System.Windows.Media.Color c, int op_min, int op_max, double freq, double phase, int type,sequenceValue seq): this("Flicker", X, Y, Width, Height, c, op_min, op_max, freq, phase, type, seq)
        {
            
        }
        /// <summary>
        /// copy a flicker
        /// </summary>
        /// <param name="f"></param>
        public Flicker(Flicker f) : this(f.Name, f.X, f.Y, f.Width, f.Height, f.color1, f.Opacity_Min, f.Opacity_Max, f.Frequency, f.Phase, (int)f.Type,f.sequence)
		{
			
		}
		/// <summary>
		/// copy a flicker
		/// </summary>
		/// <param name="f"></param>
		/// <returns></returns>
		public static Flicker copy(Flicker f)
		{
			return new Flicker(f);
		}
		public override string ToString()
		{
			return string.Format("Flicker at: {0},{1} Size: {2},{3}", X, Y, Width, Height);
		}
		
	}

    public class sequenceValue
    {
        public type Type;
        public List<sequenceValue> contained_sequence = new List<sequenceValue>();
        public CondType cond;
        public double value;
        public enum type
        {
            Block,
            Loop,
            Active,
            Inactive
        }
        public enum CondType
        {
            KeyPress = 2,
            Time = 3,
            Always = 1,
            Never = 0,
            None
        }
        private sequenceValue()
        {
            Type = type.Block;
            cond = CondType.Always;
            value = 0.0;
        }
        public sequenceValue(type t, CondType c)
        {
            Type = t;
            cond = c;
            value = 0;
        }
        public sequenceValue(type t, CondType c, double v)
        {
            Type = t;
            cond = c;
            value = v;
        }
        public sequenceValue addSeq(int pos, sequenceValue seq)
        {
            contained_sequence.Insert(pos, seq);
            return this;
        }
        public sequenceValue addSeq(sequenceValue seq)
        {
            contained_sequence.Add(seq);
            return this;

        }
        public sequenceValue addSeq(type t, CondType c)
        {
            var seq = new sequenceValue(t, c);
            addSeq(seq);
            return this;
        }
        public sequenceValue addSeq(int pos, type t, CondType c)
        {
            var seq = new sequenceValue(t, c);
            addSeq(pos, seq);
            return this;
        }
        public sequenceValue addSeq(int pos, type t, CondType c, double v)
        {
            return addSeq(pos, new sequenceValue(t, c, v));
        }
        public sequenceValue addSeq(type t, CondType c, double v)
        {
            return addSeq(new sequenceValue(t, c, v));
        }
        public void removeSeq(int pos)
        {
            contained_sequence.RemoveAt(pos);
        }
        public void removeSeq(sequenceValue seq)
        {
            contained_sequence.Remove(seq);
        }
        public override String ToString()
        {
            return Type.ToString() + ":[" + contained_sequence.ToString() + "]\n" + cond.ToString() + ":" + value + "\n";
        }
        public sequenceValue GetSequence(int pos) { return contained_sequence[pos]; }
    }
}

File: Interface2App2/Flicker.py
from enum import Enum
from PyQt5.QtGui import QColor


class FreqType(Enum):
    Sine = 1
    Root_Square = 3
    Square = 2
    Maximum_Length_Sequence = 4
    Random = 0
    Custom = 5


class SeqType(Enum):
    Block = 1
    Loop = 2
    Active = 3
    Inactive = 4


class SeqCondition(Enum):
    Never = 0
    Always = 1
    KeyPress = 2
    Time = 3


class SequenceBlock:
    def __init__(self, seq_type=SeqType.Block, cond_type=SeqCondition.Never, value=0.0,*args):
        self.seqType = seq_type
        self.Condition = cond_type
        self.value = value
        if len(args)>0:
            self.contained_sequence = args
        else:
            self.contained_sequence=[]

    def AddSeq(self, seq, index=None):
        if index==None:
            index= len(self.contained_sequence)
        self.contained_sequence.insert(index, seq)
        return self

    def removeSeq(self, seq):
        self.contained_sequence.remove(seq)

    def removeSeqByIndex(self, index):
        self.contained_sequence.pop(index)

    def __len__(self):
        return len(self.contained_sequence)

    def __repr__(self):
        return "Type: " + self.seqType.name + " Condition: " + self.Condition.name + str(self.contained_sequence) + "\n"


class Flicker:
    def __init__(self, name="Flicker", x=0, y=0, width=100, height=100, frequency=1.0, phase=0.0, type_freq=FreqType.Sine,
                 opacity_min=0, opacity_max=100, color=QColor(255, 255, 255), sequence=(SequenceBlock().AddSeq(SequenceBlock(SeqType.Active,SeqCondition.Never))), image=None,
                 IsImageFlicker=False):
        self.Name = name
        self.X = int(x)
        self.Y = int(y)
        self.Width = width
        self.Height = height
        self.Frequency = frequency
        self.Phase = phase
        self.Type = type_freq
        self.Opacity_Min = opacity_min
        self.Opacity_Max = opacity_max
        self.Color = color

        self.sequence = sequence
        self.image = image
        self.IsImageFlicker = IsImageFlicker

    def __repr__(self):
        return "Flicker X:" + str(self.X) + " Y:" + str(self.Y) + " Width" + str(self.Width) + " Height" + str(
            self.Height) + " Type:" + self.Type.name
    def __copy__(self):
        f=Flicker()
        for key in self.__dict__:
            f.__setattr__(key,self.__dict__[key])
        return f
